File: c++/Binary_Tree_Postorder_Traversal.cpp
/* Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1]. */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void posorder(TreeNode* node, vector<int>& res){
        if(node){
            if(node->left){
                posorder(node->left,res);
            }
            if(node->right){
                posorder(node->right,res);
            }
            res.push_back(node->val);
        }
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        posorder(root,res);
        return res;
    }
};

File: python/Binary_Tree_Postorder_Traversal.py
"""
Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1]. """
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def traversal(root, order):
            if not root : return
            if root.left : traversal(root.left,order)
            if root.right : traversal(root.right,order)
            order.append(root.val)
        order = []
        traversal(root, order)
        return order        
